coordinator/src/main/scala/com/socrata/datacoordinator/service/Service.scala

```diff
@@ -221,7 +224,32 @@ class Service(serviceConfig: ServiceConfig,
       "dataset" -> JString(formatDatasetId(name)),
       "schema" -> jsonifySchema(schema))
 
+  // Repeatedly tries to get a new thread by bumping numThreads, using
+  // compareAndSet to make sure our value is valid
+  @annotation.tailrec
+  private def tryGetMutationThread(numTries: Int = 100): Boolean = {
+    if (numTries <= 0) {
+      log.info("tryGetMutationThread: out of tries")
+      false
+    } else {
+      val currThreads = numThreads.get()
+      if (currThreads >= serviceConfig.maxMutationThreads) {
+        log.info(s"tryGetMutationThread: too many threads ($currThreads), waiting for some to finish")
+        Thread sleep 100
+        tryGetMutationThread(numTries - 1)
+      } else if (numThreads.compareAndSet(currThreads, currThreads + 1)) {
+        true
+      } else {
+        Thread sleep 100
+        tryGetMutationThread(numTries - 1)
+      }
+    }
+  }
+
   def withMutationScriptResults[T](f: => HttpResponse): HttpResponse = {
+    if (!tryGetMutationThread()) {
+      return err(ServiceUnavailable, "mutation.threads.maxed.out")
```

Comment:
Nit: should be "maxed-out", since I don't think we would have another value like "maxed.in".  Periods are hierarchy.